core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+  }
+}
+
+/**
+ * THe instrumentation object to be pass along the flow in the produce request
+ *
+ * @param time The object to obtain timestamp
+ * @param requiredAcks The associated required acks for the produce request
+ */
+class ProduceRequestInstrumentation(val time: Time,
+                                    val requiredAcks: Short) {
+  val marks: mutable.Map[Stage.Value, Long] = mutable.Map()
+  var topicPartitionsInRequest: Iterable[TopicPartition] = List()
+
+  // Implicitly add init stage on construction
+  markStage(ProduceRequestInstrumentation.Stage.Init)
+
+  /**
+   * The core of the instrumentation.
+   *
+   * When the object is passed along, at any point in the code flow, it can invoke this method
+   * to mark the beginning of the stage with the current timestamp.
+   * The timestamps can later be used to calculate time taken in each stage via a diff on the 2 time stamps
+   *
+   * @param stage The stage to mark for beginning
+   */
+  def markStage(stage: Stage.Value): Unit = {
+    marks.put(stage, time.milliseconds())
+  }
+}
+object ProduceRequestInstrumentationLogger {
+  private val logger = Logger("produce.request.instrumentation.logger")
+  /**
+   * The config to control the instrumentation behavior
+   *
+   * @param thresholdToLogMs If the request completed within the threshold, don't log
+   * @param logRatio         Ratio of requests to log, applied after all funnels/thresholds
+   */
+  case class Config(thresholdToLogMs: Long,
+                    logRatio: Double)
+
+  val ReconfigurableConfigs: Set[String] = Set(
+    KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp,
+    KafkaConfig.LiLongTailProduceRequestLogRatioProp,
+  )
+}
+
+/**
+ * Responsible for logging the instrumentation details.
+ * The config of the logger is associated with DynamicBrokerConfig lifecycle
+ *
+ * @param kafkaConfig For configuring dynamic broker config lifecycle
+ * @param replicaManager Used to obtain the # of partitions in PendingShrinkIsr
+ */
+class ProduceRequestInstrumentationLogger(kafkaConfig: KafkaConfig,
+                                          time: Time,
+                                          rnd: scala.util.Random,
+                                          replicaManager: ReplicaManager) extends Logging with BrokerReconfigurable {
+  override lazy val logger = ProduceRequestInstrumentationLogger.logger
+
+  // This is a cache of the dynamic config.
+  // Should be initialized/reconfigured within the BrokerReconfigurable lifecycle
+  validateReconfiguration(kafkaConfig)
+  kafkaConfig.dynamicConfig.addBrokerReconfigurable(this)
+  var config: Config = extractConfigFromKafkaConfig(kafkaConfig)
+
+  override def reconfigurableConfigs: collection.Set[String] = ProduceRequestInstrumentationLogger.ReconfigurableConfigs
+
+  override def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+    val thresholdMs = newConfig.longTailProduceRequestLogThresholdMs
+    if (thresholdMs < -1) {
```

Comment:
What is the use case for "-1"?

Reply:
Good catch.  I changed it to 0 (and the corresponding `.define(..., atLeast(0)...)`